Remove a bunch of dead code from the case data page

esoergel opened the pull request:
## Product Description
no visible changes

## Technical Summary
PRing into https://github.com/dimagi/commcare-hq/pull/34640
This is a bunch of dead code I found.  Just grabbed a thread and kept pulling.

## Feature Flag


## Safety Assurance


### Safety story


### Automated test coverage


### QA Plan


### Rollback instructions
- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

esoergel committed:
related_type_info was removed in 5befebf3a09

esoergel committed:
Remove type info from get_case_hierarchy

esoergel committed:
Remove type_info from sortkey

esoergel committed:
Remove type_info from process_case_hierarchy

esoergel committed:
descendant_types was removed a decade ago

in 811dff3807f

esoergel committed:
This is all gone now too

case.add_child_data is gone, and edit_form_id and description don't exist in
edit_data anymore

esoergel committed:
Remove useless span

According to my dev tools:

> width has no effect on this element since it has a display of inline

And removing it changes nothing

esoergel committed:
Remove unused touchforms utils

The last usage of these was removed in 202201bb511

esoergel committed:
This is always True

esoergel committed:
lint

esoergel committed:
Just link the URL the normal way

esoergel committed:
We're already looping through parents

esoergel committed:
Now this doesn't actually do anything

esoergel committed:
We only ever use 'case_list' from here

esoergel committed:
Rework recursive logic to be a bit simpler

millerdev commented on corehq/apps/reports/view_helpers.py:
```diff
@@ -23,15 +23,13 @@ def case_hierarchy_context(case, timezone=None):
-            if parent_case:
-                last_parent_id = parent_case.case_id
-            else:
                 last_parent_id = None
+            else:
+                parent_case.index_info = _index_to_context(idx, is_ancestor=True)
+                parent_cases.append(parent_case)
+                last_parent_id = parent_case.case_id
```
nit: I think this could be simplified a bit more
```suggestion
            else:
                parent_case.index_info = _index_to_context(idx, is_ancestor=True)
                parent_cases.append(parent_case)
        last_parent_id = parent_cases[-1].case_id if parent_cases[-1] else None
```

esoergel replied:
Thanks for the feedback - looking at this a little more closely, it seems this is actually only necessary for the main case anyways, as the rest will have it set in `get_case_hierarchy`.  This gets simpler and more self-explanatory if we modify only the main case.

I'm also kinda tempted to set it inside the `else` block here, so it gets set for each existing parent, with the last one winning.  If we even do have missing parents, that seems like the best behavior anyways.  That said, I'll avoid the behavior change for now.  Just pushed a commit.

millerdev commented on corehq/apps/reports/view_helpers.py:
```diff
@@ -23,15 +23,13 @@ def case_hierarchy_context(case, timezone=None):
+                parent_cases.append(parent_case)
+                last_parent_id = parent_case.case_id
 
         for c in descendant_case_list:
```
This whole loop could be skipped if `parent_cases[-1] is None`, and if that is done then `last_parent_id = parent_cases[-1].case_id` could be moved inside the `if` block and removed from the last line of the suggestion above.

millerdev reviewed (commented):
I think this looks good, but had a hard time evaluating the safety of some of the refactors. Going to drop this review with my few comments, and hopefully someone who is more familiar with this code can help you get it approved.

orangejenny committed:
Merge branch 'master' into es/case-data-dead-code

orangejenny commented on corehq/apps/reports/view_helpers.py:
```diff
@@ -188,29 +186,21 @@ def process_output(case_output, depth=0):
 
-def get_case_hierarchy(case, type_info):
-    def get_children(case, referenced_type=None, seen=None):
-        seen = seen or set()
```
agreed

orangejenny commented on corehq/apps/reports/view_helpers.py:
```diff
@@ -76,52 +76,12 @@ def _index_to_context(index, is_ancestor):
     }
 
 
-def normalize_date(val):
```
You also need to remove https://github.com/dimagi/commcare-hq/blob/master/corehq/ex-submodules/casexml/apps/case/tests/test_tags.py, right?

esoergel replied:
oh yes, good find, thanks, must've forgotten about that.  Hoping that's the cause of the build failures, since they pass now locally.

orangejenny commented on corehq/ex-submodules/casexml/apps/case/templates/case/partials/case_hierarchy.html:
```diff
@@ -42,7 +42,8 @@
             <div class="pull-right btn-group">
-              <a class="btn btn-sm btn-default view-related-case-link" href="{{ case.edit_data.view_url }}">
+              <a class="btn btn-sm btn-default view-related-case-link"
+                 href="{% url 'case_data' domain case.case_id %}">
```
normal indeed

orangejenny reviewed (commented):
Looks good. Test failures are suspicious, though.

esoergel committed:
PR feedback

esoergel committed:
only the present case needs a parent set here

everything else in descendant_case_list should get it set inside
get_case_hierarchy

orangejenny commented on corehq/apps/reports/view_helpers.py:
```diff
@@ -25,15 +25,11 @@ def case_hierarchy_context(case, timezone=None):
-        for c in descendant_case_list:
-            if not getattr(c, 'treetable_parent_node_id', None) and last_parent_id:
-                c.treetable_parent_node_id = last_parent_id
+        case.treetable_parent_node_id = parent_cases[-1].case_id if parent_cases[-1] else None
```
Won't `parent_cases[-1]` fail if `parent_cases` is empty?

esoergel replied:
`parent_cases` can't be empty - this is all inside a `if case.live_indices`, and something is added to `parent_cases` for every live index.  It does feel like there's still room to make this whole thing clearer though, and I'm open to suggestions if you've got an approach you like.

orangejenny replied:
ah, this makes sense. I misread the indentation, I raised this because I thought it was outside of the if.

millerdev commented on corehq/apps/reports/view_helpers.py:
```diff
@@ -74,7 +74,8 @@ def _sortkey(child):
 def get_case_hierarchy(case):
-    def get_children(case, seen):
+    def get_children(case, seen=None):
+        seen = seen or set()
```
I didn't notice this until re-review, and your original change would have fixed the issue, but I think this may have a bug in it. If the passed in set is empty a new set will be created in a recursive call, which may then add items that should but will not be seen by the outer call.
```suggestion
        if seen is None:
            seen = set()
```
~The way it was written before with a mutable default `def get_children(case, seen=set()):` is syntactically simpler and would work as well, although I'm guessing you think that smells.~ Well nevermind. It appears this bug was also in the original code. Not sure what made me think there was a mutable default.

esoergel replied:
Oh good point, I hadn't noticed that.  I think I'll actually change it back to the way I had it before, which I find more explicit - we know exactly how this is going to be used, so why fork the logic?  I'd agree with you and Jenny  if this function were a utility or public interface, but it's an internal function used for a recursive loop.  I don't think the function declaration and invocation are separate in the same way normal functions should be.

esoergel committed:
move var init back to function call
